jsonrpc/debug_endpoint.go

```diff
@@ -0,0 +1,231 @@
+}
+
+func (d *Debug) TraceBlockByHash(
+	blockHash types.Hash,
+	config *TraceConfig,
+) (interface{}, error) {
+	block, ok := d.store.GetBlockByHash(blockHash, true)
+	if !ok {
+		return nil, fmt.Errorf("block %s not found", blockHash)
+	}
+
+	return d.traceBlock(block, config)
+}
+
+func (d *Debug) TraceBlock(
+	input string,
+	config *TraceConfig,
+) (interface{}, error) {
+	blockByte, decodeErr := hex.DecodeHex(input)
+	if decodeErr != nil {
+		return nil, fmt.Errorf("unable to decode block, %w", decodeErr)
+	}
+
+	block := &types.Block{}
+	if err := block.UnmarshalRLP(blockByte); err != nil {
+		return nil, err
+	}
+
+	return d.traceBlock(block, config)
+}
+
+func (d *Debug) TraceTransaction(
+	txHash types.Hash,
+	config *TraceConfig,
+) (interface{}, error) {
+	tx, block := GetTxAndBlockByTxHash(txHash, d.store)
+	if tx == nil {
+		return nil, fmt.Errorf("tx %s not found", txHash.String())
+	}
+
+	if block.Number() == 0 {
+		return nil, ErrTraceGenesisBlock
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	tracer, err := newTracer(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	return d.store.TraceTxn(block, tx.Hash, tracer)
+}
+
+func (d *Debug) TraceCall(
+	arg *txnArgs,
+	filter BlockNumberOrHash,
+	config *TraceConfig,
+) (interface{}, error) {
+	header, err := GetHeaderFromBlockNumberOrHash(filter, d.store)
+	if err != nil {
+		return nil, ErrHeaderNotFound
+	}
+
+	tx, err := DecodeTxn(arg, d.store)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the caller didn't supply the gas limit in the message, then we set it to maximum possible => block gas limit
+	if tx.Gas == 0 {
+		tx.Gas = header.GasLimit
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	tracer, err := newTracer(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	return d.store.TraceCall(tx, header, tracer)
+}
+
+func (d *Debug) traceBlock(
+	block *types.Block,
+	config *TraceConfig,
+) (interface{}, error) {
+	if block.Number() == 0 {
+		return nil, ErrTraceGenesisBlock
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	tracer, err := newTracer(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	return d.store.TraceBlock(block, tracer)
+}
+
+// newTracer creates new tracer by config
+func newTracer(
+	ctx context.Context,
+	config *TraceConfig,
+) (tracer.Tracer, error) {
+	var (
+		timeout = defaultTraceTimeout
+		err     error
+	)
+
+	if config.Timeout != nil {
+		if timeout, err = time.ParseDuration(*config.Timeout); err != nil {
+			return nil, err
+		}
+	}
+
+	tracer := structtracer.NewStructTracer(structtracer.Config{
+		EnableMemory:     config.EnableMemory,
+		EnableStack:      !config.DisableStack,
+		EnableStorage:    !config.DisableStorage,
+		EnableReturnData: config.EnableReturnData,
+	})
+
+	// cancellation of context is done by caller
+	timeoutCtx, _ := context.WithTimeout(context.Background(), timeout)
```

Comment:
You can't call cancel func inside this function because the process doesn't begin inside. The callee must call it but I added context instead of this but it wasn't used. I make this func return cancel func

Reply:
3bf99bb2067186017e8f35cbc4fa85ad061a5579